src/tools/rbd_mirror/image_map/Policy.cc

```diff
@@ -0,0 +1,239 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "common/debug.h"
+#include "common/errno.h"
+
+#include "librbd/Utils.h"
+#include "Policy.h"
+
+#define dout_context g_ceph_context
+#define dout_subsys ceph_subsys_rbd_mirror
+#undef dout_prefix
+#define dout_prefix *_dout << "rbd::mirror::image_map::Policy: " << this \
+                           << " " << __func__
+
+namespace rbd {
+namespace mirror {
+namespace image_map {
+
+using librbd::util::unique_lock_name;
+
+namespace {
+
+bool can_shuffle_image(utime_t now, utime_t last_shuffled_time, int migration_throttle) {
+  return !(migration_throttle > 0 && (now - last_shuffled_time) < migration_throttle);
```

Comment:
Related to the comment about integrating ```ImageAction``` and ```Policy``` -- we shouldn't be migrating the image if that image already has an action in-progress. The only thing that should override an in-progress action is when the peer for the in-progress destination action is removed (i.e. it's safe to remap).

Reply:
checking back on this one -- for the instance removal case, we would want to restart the (pending) actions after the images (currently mapped to the removed instance) have been migrated to another mirror peer, right?

Reply:
When the policy is informed that a peer has been removed, you don't need to tell that peer that it should remove an image (since the peer is dead / blacklisted). You just need to map it somewhere else. Similarly, if you have an image that is in-progress being mapped to a removed peer, you would need to pick a new peer to map it to.